### PR TITLE
PoC on new currying

### PR DESCRIPTION
**Pull request by nlepage**
**Issue :** #297

### CODE REVIEW

**Comment by nlepage** [comment]
@frinyvonnick Ready to review

File: packages/immutadot/src/core/curry.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Manage a curried call.
+ * @returns {function|*} Either a new function or the result of the call to the actual function
+ * @this Object
+ */
+function curried(...args) {
+  const { fn, last, minArity } = this
+
+  if (last) {
+    const obj = args.length === 0 ? undefined : args[0]
```
Comment (EmrysMyrddin) [inline packages/immutadot/src/core/curry.js]: Why do you check for the length of the array ?

```es6
const [ obj ] = args
```

This doesn't achieve the same result ?

**Comment by EmrysMyrddin** [comment]
After some thinking, perhaps it is possible to have only one function `curry` that doesn't require the tow implementations `curried` and `myabeCurried` and that doesn't rely on `this`.

Something like this (haven't tested it) :
```es6
export function curry(fn, minArity = fn.length) {
  const curried = previousArgs => (...args) => {
    const allArgs = previousArgs.concat(args)
 
    if (allArgs.length >= minArity - 1) {
      const obj = this.args[allArgs.length - 1)
      return fn(obj, ...allArgs.slice(0, -1))
    }
 
    return curried(allArgs)
  }

  return curried([])
}
```

**Comment by nlepage** [comment]
@EmrysMyrddin What you're suggesting doesn't handle a non curried call, it always takes the last parameter for `obj`.

File: packages/immutadot/src/core/curry.js (hunk 1 above)
Comment (nlepage) [reply packages/immutadot/src/core/curry.js]: Yes, somehow I forgot how "flexible" JavaScript is...

**Comment by nlepage** [comment]
I like the usage of `bind` and `this` as a context, this doesn't need any dynamic function creations.

I had created the `maybeCurried` in order to separate concerns:
 - `maybeCurried` chooses between curried and non curried
 - `curried` handles curried calls

I agree that mutating the context is not a good idea at all.
The first time I had written it with destructuring between each `bind`, I don't know why I changed it, I'm going to revert it...

**Comment by EmrysMyrddin** [comment]
For me, the `bind` function generate a new function when called. If it was not the case, calling `bind` on a function could have unexpected behavior.

File: packages/immutadot/src/core/curry.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+function curried(...newArgs) {
+  const { fn, last, minArity } = this
+
+  if (last) {
+    const [obj] = newArgs
+    return fn(obj, ...this.args)
+  }
+
+  const args = this.args.concat(newArgs)
```
Comment (frinyvonnick) [inline packages/immutadot/src/core/curry.js]: Why are you doing this on a separate line ?

Comment (nlepage) [reply packages/immutadot/src/core/curry.js]: `args` is used 2 times in the destructuring

**Comment by nlepage** [comment]
I read `bind`'s documentation, it clearly states that a new function is created.
I tried to make some performance tests: https://jsperf.com/binded-vs-curried/1
These doesn't seem very reliable:
 - On Chrome bind is ~2% slower
 - On Firefox curried is 10x faster ?!

I find Firefox's result hard to believe, I suspect it is doing some shady optimization like "hey, he doesn't use the value, let's not compute it" and it can't do the same with bind...
That's why I tried to send the value to a `noop` function, but it doesn't change anything.

Let me know what's the result on your computers.

Anyway I'm leaning towards rewriting without using `bind`.

**Comment by frinyvonnick** [comment]
`bind` is 11% slower on Chrome for me 😄

**Comment by EmrysMyrddin** [comment]
I have the same results : bind is 4% slower on chrome and 91% slower on firefox.

**Comment by EmrysMyrddin** [comment]
Ok finaly I have tryed to do it with only one function. Not sure if the result is better. At least it doesn't duplicate some behavior

```es6
export function curry(fn, minArity = fn.length) {
  const curried = (previousArgs = [], last = false) => (...args) => {
    // Short circuit if no curry is necessary
    if (previousArgs.length === 0 && args.length >= minArity) return fn(...args)

    if (last) {
      const [obj] = args
      return fn(obj, ...previousArgs)
    }

    const allArgs = previousArgs.concat(args)
    const nextLast = allArgs.length >= minArity - 1

    return curried(allArgs, nextLast)
  }

  return curried()
}
```

**Comment by nlepage** [comment]
OK, one lined it !
Just kidding... 2 lined it.

Usually I avoid using ternary in this kind of function.
What do you think ? Should I use if returns instead ?

**Comment by EmrysMyrddin** [comment]
👍 Very nice !

**Comment by frinyvonnick** [comment]
I find it hard to read !

**Pull request merged**